File: name.js
var letters = 'abcdefghijklmnopqrstuvwxyz';
var onAddNameClick = function() {
	addNameToList(makeNameFromUI());
};
var getRadioValue = function(name) {
	var radios = document.getElementsByName(name);
	if(radios.length == 0) throw 'No elements with name: ' + name;

	for (var i = 0, length = radios.length; i < length; i++) {
		if (radios[i].checked) {
			return radios[i].value;
		}
	}
	throw 'No radio button with name ' + name + ' are checked';
};
var makeNameFromUI = function() {
	var sex = getRadioValue('sex');
	var length_type = getRadioValue('length_type');
	var length;
	if(length_type === 'length_exact') {
		length = parseInt(document.getElementById('length').value,10);
	} else {
		var length_min = parseInt(document.getElementById('length_range_min').value, 10);
		var length_max = parseInt(document.getElementById('length_range_max').value, 10);
		if(length_max < length_min) {
			length_max = length_min;
		}
		length = getRandomInt(length_min, length_max);
	}
	return makeNames(stats, length, 1, sex);
};
var makeNames = function(stats, len, count, gender) {
	if(count === undefined) count = 1;
	var un = null;
	if(gender === 'n') {
		un = stats.ngram;
	} else if(gender[0] === 'm') {
		un = stats.ngram_m;
	} else {
		un = stats.ngram_f;
	}
	var names = [];
	for(var i = 0; i < count; i++) {
		names.push(onename(un, len))
	}
	return names;
};
// Returns a random integer between min and max
var getRandomInt = function(min, max) {
	return Math.floor(Math.random() * (max - min + 1) + min);
};
// sample from data with probabilities specified by prob
// prob neet not sum to 1 but must be non-negative values
var sample = function(data, prob) {
	if(prob !== undefined && data.length != prob.length) {
		throw "Data (" + data.length + ") and probability weights (" + prob.length + ") must be same length";
	}
	if(prob === undefined) {
		return data[getRandomInt(0, data.length)];
	}
	var cum = [prob[0]];
	for(var i = 1; i < data.length; i++) {
		cum.push(cum[i-1] + prob[i]);
	}
	var r = Math.random() * cum[cum.length - 1];
	var index = 0;
	for(index = 0; cum[index] < r; index++) {
	}
	return data[index];
};
var onename = function(ngram, len) {
	if(len === undefined) {
		// todo magic numbers to define range of name lengths
		len = getRandomInt(4,8);
	}
	// generate first letter based on frequency
	var name = sample(letters, ngram.gram);
	len--;
	// generate second letter based on previous unigram
	name = name.concat(sample(letters, ngram[name[name.length - 1]].gram));
	len--;
	// generate remaining letters based on previous bigrams
	while(len > 0) {
		name = name.concat(sample(letters, ngram[name[name.length - 2]][name[name.length - 1]].gram))
		len--;
	}
	return name;
};
var addNameToList = function(name) {
	// create the dom element
	var element = document.createElement('div');
	// add text
	element.appendChild(document.createTextNode(name));
	document.getElementById('name_output').appendChild(element);
};

File: name.R
library('rjson')

vowels <- c('a','e','i','o','u','y')
consonants <- c('b','c','d','f','g','h','j','k','l','m','n','p','q','r','s','t','v','w','x','z')

learn <- function(file, limit = 0, weighted = TRUE) {
	if(missing(file)) {
		file <- 'names/yob2012.txt'
	}
	res <- list()
	# get name list
	namelist <- read.table(file, sep=',', header=FALSE)
	names(namelist) <- c('name', 'gender', 'count')
	namelist$name <- tolower(as.character(namelist$name))
	namelist.f <- subset(namelist, gender == 'F' & count > limit)
	namelist.m <- subset(namelist, gender == 'M' & count > limit)
	res$ngram.f <- aggregate(namelist.f, weighted)
	res$ngram.m <- aggregate(namelist.m, weighted)
	res$ngram <- list()
	res$ngram$gram <- res$ngram.m$gram + res$ngram.f$gram
	for(l in letters) {
		res$ngram[[l]]$gram <- res$ngram.m[[l]]$gram + res$ngram.f[[l]]$gram
		for(d in letters) {
			res$ngram[[l]][[d]]$gram <- res$ngram.m[[l]][[d]]$gram + res$ngram.f[[l]][[d]]$gram
			res$ngram[[l]][[d]]$last <- res$ngram.m[[l]][[d]]$last + res$ngram.f[[l]][[d]]$last
		}
	}
	res
}
aggregate <- function(namelist, weighted) {
	# initialize unigram list
	letterlist <- sample(0,26,rep=TRUE)
	names(letterlist) <- letters
	ngram <- list()
	ngram$gram <- letterlist
	for(l in letters) {
		ngram[[l]] <- list()
		ngram[[l]]$gram <- letterlist
		for(d in letters) {
			ngram[[l]][[d]]$gram <- letterlist
			ngram[[l]][[d]]$last <- letterlist
		}
	}
	# get statistics
	for(i in seq(namelist$name)) {
		if(weighted) {
			weight <- namelist[i,3] / 100;
		} else {
			weight <- 1
		}
		split <- strsplit(namelist[i,1], '')[[1]]
		# increment frequency of first letter
		ngram$gram[[split[1]]] <- ngram$gram[[split[1]]] + 1
		# increment frequency of letter based on previous unigram
		ngram[[split[1]]]$gram[[split[2]]] <- ngram[[split[1]]]$gram[[split[2]]] + 1
		for(j in 2:length(split)) {
			if(j > 2) {
				ngram[[split[j - 2]]][[split[j - 1]]]$gram[[split[j]]] <- ngram[[split[j - 2]]][[split[j - 1]]]$gram[[split[j]]] + weight
				if(j == length(split) - 1) {
					ngram[[split[j - 2]]][[split[j - 1]]]$last[[split[j]]] <- ngram[[split[j - 2]]][[split[j - 1]]]$last[[split[j]]] + weight
				}
			}
		}
	}
	ngram
}

name <- function(stats, len = FALSE, count = 1, gender = 'both') {
	if(gender == 'both') {
		# TODO combine data
		un <- stats$ngram
	} else if(head(gender,1) == 'm') {
		un <- stats$ngram.m
	} else {
		un <- stats$ngram.f
	}
	names <- c()
	for(i in seq(count)) {
		names <- append(names, onename(un, len))
	}
	names
}
onename <- function(ngram, len = FALSE) {
	if(len == FALSE) {
		# todo magic numbers to define range of name lengths
		len <- sample(4:8, 1)
	}
	# generate first letter based on frequency
	name <- sample(letters, 1, prob = ngram$gram)
	len <- len - 1
	# generate second letter based on previous unigram
	# TODO train these only on first letter unigrams?
	name <- append(name, sample(letters, 1, prob = ngram[[tail(name, 1)]]$gram))
	len <- len - 1
	# generate remaining letters based on previous bigrams
	while(len > 0) {
		name <- append(name, sample(letters, 1, prob = ngram[[head(tail(name, 2), 1)]][[tail(name, 1)]]$gram))
		len <- len - 1
	}
	paste(name, collapse = '')
}
writeStats <- function(stats, file, zeros = 13) {
	stats$f <- vectorify(stats$ngram.f, zeros)
	stats$ngram.f <- NULL
	stats$m <- vectorify(stats$ngram.m, zeros)
	stats$ngram.m <- NULL
	stats$n <- vectorify(stats$ngram, zeros)
	stats$ngram <- NULL
	write(toJSON(stats), file)
}
vectorify <- function(ngram, zeros) {
	# convert frequency lists to vectors and json
	ngram$F <- as.vector(ngram$gram)
	ngram$gram <- NULL
	for(l in letters) {
		ngram[[l]]$F <- as.vector(ngram[[l]]$gram)
		ngram[[l]]$gram <- NULL
		for(d in letters) {
			ngram[[l]][[d]]$F <- reduceFreq(ngram[[l]][[d]]$gram,zeros)
			ngram[[l]][[d]]$gram <- NULL
			ngram[[l]][[d]]$L <- reduceFreq(ngram[[l]][[d]]$last,zeros)
			ngram[[l]][[d]]$last <- NULL
		}
	}
	ngram
}
reduceFreq <- function(freq,zeros) {
	# if there are enough zeros, store a map of letters and their non-zero frequencies instead of the array
	t <- table(freq)
	if(!'0' %in% names(t) || t['0'] < zeros) {
		res <- as.vector(freq)
	} else {
		# remove zeros from list
		res <- freq[-which(freq == 0)]
	}
	res
}